### PR TITLE
[Collab] History dashboard

### PR DESCRIPTION
**Pull request by jeffrey-jian**
## Description

This PR introduces the new `/history` frontend page, where users can view their past collaboration sessions.
Screenshot 2024-11-05 at 5 38 48 PM

`HistoryTable` is filterable by 
- question title
- question category(s)
- question complexity(s)

`HistoryTable` is sortable by
- started_at datetime
- ended_at datetime

Note: not sortable by question and user-related fields as we can't easily do `JOIN` operations on the different tables.

### Potential Enhancements 
@jonyeokj @gongg21 
- Prettier date string

#### Frontend:
- Added `HistoryTable` and related components / defs.

#### Backend:
- Added support for retrieval , sorting, filtering, pagination for data in `HistoryTable`.

### Next steps
- [ ] Add `/history/:id` page, which displays the question and code.

### CODE REVIEW

File: project/apps/collaboration-service/src/adapters/db/collaboration.supabase.ts [hunk 1]
```diff
@@ -93,21 +99,49 @@ export class CollaborationSupabase implements CollaborationRepository {
...
       .from(this.COLLABORATION_TABLE)
       .select('*', { count: 'exact' });
 
+    if (filteredQuestionIds) {
+      queryBuilder = queryBuilder.in('question_id', filteredQuestionIds);
+    }
+
     // Apply user filtering
     if (user_id && collab_user_id) {
       // Both user_id and collab_user_id are provided
```
Comment (firwer) [inline project/apps/collaboration-service/src/adapters/db/collaboration.supabase.ts:147]: Optional: should we consider allowing search by username of collab peer on the history given that this is currently implemented as a filter field?

If difficult to implement, perhaps can remove this

Comment (jeffrey-jian) [reply project/apps/collaboration-service/src/adapters/db/collaboration.supabase.ts:147]: Not the most straightforward to implement, bcos we cant easily join the `profile` and `collaboration` tables.
Will do this in the future if there time 👍

File: project/apps/web/components/history/history-table/HistoryTable.tsx [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+import { QUERY_KEYS } from '@/constants/queryKeys';
+import useDebounce from '@/hooks/useDebounce';
+import { useMe } from '@/hooks/useMe';
+import { fetchCollabs } from '@/lib/api/collab';
+import { useQuestionsStore } from '@/stores/useQuestionStore';
+
+import { columns } from './columns';
+import { HistoryTableToolbar } from './HistoryTableToolbar';
+
+export type CollabInfoiWithPartner = CollabInfoDto & {
```
Comment (firwer) [inline project/apps/web/components/history/history-table/HistoryTable.tsx:35]: nit: typo

File: project/apps/web/components/history/history-table/HistoryTable.tsx [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  const dataWithPartner = data?.collaborations.map((collab) => {
+    const partner =
+      collab.collab_user1.id === userData!.id
+        ? collab.collab_user2
+        : collab.collab_user1;
+    return {
+      ...collab,
+      partner,
+    };
+  }) as CollabInfoiWithPartner[];
```
Comment (firwer) [inline project/apps/web/components/history/history-table/HistoryTable.tsx:154]: nit: same as earlier

File: project/apps/web/components/history/history-table/columns.tsx [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+import { ColumnDef } from '@tanstack/react-table';
+import Link from 'next/link';
+
+import { DataTableColumnHeader } from '@/components/data-table/DataTableColumnHeader';
+import DifficultyBadge from '@/components/DifficultyBadge';
+import { Badge } from '@/components/ui/badge';
+
+import { CollabInfoiWithPartner } from './HistoryTable';
+
+export const columns: ColumnDef<CollabInfoiWithPartner>[] = [
```
Comment (firwer) [inline project/apps/web/components/history/history-table/columns.tsx:13]: nit: here as well

File: project/apps/web/components/history/history-table/HistoryTable.tsx [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+  const onColumnFiltersChange = (updater: Updater<ColumnFiltersState>) => {
+    startTransition(() => {
+      setColumnFilters(updater);
+      resetPagination();
+    });
+  };
+
+  //  dataWithPartner includes a new 'partner' field, which is the other user in the collaboration
+  const dataWithPartner = data?.collaborations.map((collab) => {
```
Comment (firwer) [inline project/apps/web/components/history/history-table/HistoryTable.tsx:145]: is it possible/suitable to memoize this perhaps? 
```
const dataWithPartner = useMemo(() => {
  if (!data || !userData) return [];
  return data.collaborations.map((collab) => {
    const partner = collab.collab_user1.id === userData.id
      ? collab.collab_user2
      : collab.collab_user1;
    return { ...collab, partner };
  }) as CollabInfoWithPartner[];
}, [data, userData]);

```

**Review by firwer — APPROVED** [review]
lgtm 👍🏻, a few small nits

File: project/apps/web/components/history/history-table/HistoryTable.tsx (hunk 5 above)
Comment (jeffrey-jian) [reply project/apps/web/components/history/history-table/HistoryTable.tsx:145]: thx for the suggestion 👍

**Pull request merged**